npm/networkPolicyController.go

```diff
@@ -0,0 +1,492 @@
+	if !ok {
+		return key, fmt.Errorf("cannot cast obj (%v) to network policy obj", obj)
+	}
+
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		return key, fmt.Errorf("error due to %s", err)
+	}
+
+	return key, nil
+}
+
+func (c *networkPolicyController) addNetworkPolicy(obj interface{}) {
+	key, err := c.needSync(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	c.workqueue.Add(key)
+}
+
+func (c *networkPolicyController) updateNetworkPolicy(old, new interface{}) {
+	key, err := c.needSync(new)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	// needSync already checked validation of casting new network policy.
+	newNetPol, _ := new.(*networkingv1.NetworkPolicy)
+	oldNetPol, ok := old.(*networkingv1.NetworkPolicy)
+	if ok {
+		if oldNetPol.ResourceVersion == newNetPol.ResourceVersion {
+			// Periodic resync will send update events for all known network plicies.
+			// Two different versions of the same network policy will always have different RVs.
+			return
+		}
+	}
+
+	netPolCachedKey := util.GetNSNameWithPrefix(key)
+	c.npMgr.Lock()
+	cachedNetPolObj, netPolExists := c.npMgr.RawNpMap[netPolCachedKey]
+	c.npMgr.Unlock()
+	if netPolExists {
+		// if network policy does not have different states against lastly applied states stored in cachedNetPolObj,
+		// netPolController does not need to reconcile this update.
+		// in this updateNetworkPolicy event, newNetPol was updated with states which netPolController does not need to reconcile.
+		if isSameNetworkPolicy(cachedNetPolObj, newNetPol) {
+			return
+		}
+	}
+
+	c.workqueue.Add(key)
+}
+
+func (c *networkPolicyController) deleteNetworkPolicy(obj interface{}) {
+	_, ok := obj.(*networkingv1.NetworkPolicy)
+	// DeleteFunc gets the final state of the resource (if it is known).
+	// Otherwise, it gets an object of type DeletedFinalStateUnknown.
+	// This can happen if the watch is closed and misses the delete event and
+	// the controller doesn't notice the deletion until the subsequent re-list
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			metrics.SendErrorLogAndMetric(util.NetpolID, "[NETPOL DELETE EVENT] Received unexpected object type: %v", obj)
+			utilruntime.HandleError(fmt.Errorf("error decoding object, invalid type"))
+			return
+		}
+
+		if _, ok = tombstone.Obj.(*networkingv1.NetworkPolicy); !ok {
+			metrics.SendErrorLogAndMetric(util.NetpolID, "[NETPOL DELETE EVENT] Received unexpected object type (error decoding object tombstone, invalid type): %v", obj)
+			utilruntime.HandleError(fmt.Errorf("error decoding object tombstone, invalid type"))
+			return
+		}
+	}
+
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	netPolCachedKey := util.GetNSNameWithPrefix(key)
+
+	// (TODO): need to decouple this lock from npMgr if possible
+	c.npMgr.Lock()
```

Comment:
Shouldnt we take a ReadLock? Same for above

Reply:
Good catch. We can use readlock here, but current NPM code base is using global Mutex. When we decouple global Mutex from our code, we can take your suggestion. It will be different PR.